drivers/flash/flash_ll_nrfxlib.c

```diff
@@ -0,0 +1,313 @@
+	 * bytes_to_copy          : MIN(1, 2) => 1
+	 */
+	size_t max_offset = MAX(offset_32(dst), offset_32(src));
+	size_t remaining_bytes_in_word = sizeof(u32_t) - max_offset;
+	size_t bytes_to_copy = MIN(len, remaining_bytes_in_word);
+
+	/* nRF52832 Product specification:
+	 *  Only full 32-bit words can be written to Flash using the
+	 *  NVMC interface. To write less than 32 bits to Flash, write
+	 *  the data as a word, and set all the bits that should remain
+	 *  unchanged in the word to '1'. */
+	*word_dst = ~0;
+	memcpy(&((u8_t *)word_dst)[offset_32(dst)], src, bytes_to_copy);
+	return bytes_to_copy;
+}
+
+static int flash_op_write(void)
+{
+	if (is_aligned_32(flash_state.addr) &&
+	    is_aligned_32((off_t) flash_state.data) &&
+	    flash_state.len >= sizeof(u32_t)) {
+		flash_state.prev_len = MIN(align_32(flash_state.len), NRF_FICR->CODEPAGESIZE);
+		return ble_controller_flash_write((u32_t) flash_state.addr,
+						  flash_state.data,
+						  bytes_to_words(flash_state.prev_len),
+						  flash_operation_complete_callback);
+	} else {
+		flash_state.prev_len = unaligned_word_copy(&flash_state.tmp_word,
+							   (void *)flash_state.addr,
+							   flash_state.data,
+							   flash_state.len);
+		return ble_controller_flash_write((u32_t) align_32(flash_state.addr),
+						  &flash_state.tmp_word,
+						  1,
+						  flash_operation_complete_callback);
+	}
+}
+
+static int flash_op_execute(void)
+{
+	int err;
+
+	err = MULTITHREADING_LOCK_ACQUIRE();
+	if (!err) {
+		if (flash_state.op == FLASH_OP_WRITE) {
+			err = flash_op_write();
+		} else if (flash_state.op == FLASH_OP_ERASE) {
+			flash_state.prev_len = NRF_FICR->CODEPAGESIZE;
+			err = ble_controller_flash_page_erase((u32_t) flash_state.addr,
+							      flash_operation_complete_callback);
+		} else {
+			__ASSERT(0, "Unsupported operation\n");
+			err = -EINVAL;
+		}
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	return err;
+}
+
+/* Driver API. */
+
+static int btctlr_flash_read(struct device *dev, off_t offset, void *data, size_t len)
+{
+	int err;
+
+	if (!is_addr_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	err = k_sem_take(&flash_state.sem, K_FOREVER);
+	if (err) {
+		return err;
+	}
+
+	memcpy(data, (void *)offset, len);
+	k_sem_give(&flash_state.sem);
+
+	return err;
+}
+
+static int btctlr_flash_write(struct device *dev, off_t offset, const void *data, size_t len)
+{
+	int err;
+
+	if (!is_addr_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	err = k_sem_take(&flash_state.sem, K_FOREVER);
+	if (err) {
+		return err;
+	}
+
+	__ASSERT_NO_MSG(flash_state.op == FLASH_OP_NONE);
+	flash_state.op = FLASH_OP_WRITE;
+	flash_state.data = data;
+	flash_state.addr = offset;
+	flash_state.len = len;
+
+	err = flash_op_execute();
+	if (err) {
+		k_sem_give(&flash_state.sem);
+	}
+
+	return err;
+}
+
+static int btctlr_flash_erase(struct device *dev, off_t offset, size_t len)
+{
+	int err;
+	int page_count = len / NRF_FICR->CODEPAGESIZE;
```

Comment:
For anything below the ```CODEPAGESIZE``` we get 0 here - is it intentional?

Reply:
Yes, you cannot erase anything less than a full page.

Reply:
And if the ```len``` value is ```CODEPAGE+1``` we wish to erase one or two codepages? If there is any requirement for the value of ```len``` I would expect assertion here - assertion is very good documentation :)

Reply:
The API returns an error in the if-test below if `len` is not page-aligned. If C99 is acceptable, I can move the declaration and assignment below the if-test as `page_count` isn't used there.

Reply:
As far as I know, it is acceptable. And this way it would be a little easier to see what is going on.